lib/srv/reexec_linux.go

```diff
@@ -18,10 +18,47 @@
 package srv
 
 import (
+	"os"
 	"os/exec"
 	"syscall"
 )
 
+// runningInQemuUser tries to determine if we're running inside of qemu-user;
+// will err to the side of "not qemu" in case of internal errors, as that's not
+// very common - and errors while we try to figure this out have a higher chance
+// of happening on real systems anyway.
+func runningInQemuUser() bool {
+	// under qemu-user this is going to be a temporary file filled right before
+	// returning from the syscall...
+	statFile, err := os.Open("/proc/self/stat")
+	if err != nil {
+		return false
+	}
+	defer statFile.Close()
+
+	// ...and regular files with content (/proc/self/stat is not empty) report a
+	// nonzero size after fstat, whereas the actual procfs file reports zero
+	statInfo, err := statFile.Stat()
+	if err != nil {
+		return false
+	}
+
+	return statInfo.Size() != 0
```

Comment:
how did you figure out this check, is there a link somewhere? would it be possible to have a more general solution that checks the status of `/proc/self/exe`?

Reply:
This is a bit of a doozy: the reason why we can't just execve `/proc/self/exe` is that qemu-user doesn't have that as a special case in [`linux-user/syscall.c`](https://gitlab.com/qemu-project/qemu/-/blob/v6.2.0/linux-user/syscall.c); however, `readlink`/`readlinkat` and `openat` have special cases (grep for `is_proc_myself` in that file).

Opening `/proc/self/exe` is intended to give you a file descriptor on the original binary (by reopening it, sadly) but because of [the bug I linked](https://gitlab.com/qemu-project/qemu/-/issues/927) you actually get a completely bogus file descriptor back if qemu-user was launched by the `binfmt_misc` machinery, as you just get the file descriptor that was used to load the program and was then closed afterwards. If this worked we could rework our code a bit to open `/proc/self/exe` and then launch `/proc/self/fd/number` instead; sadly, that's not the case.

Because of the way the `is_proc_myself` check is implemented (and has been implemented for the past eight years or so) we can actually access the original files by going through `/proc/self/./` or `/proc/./self/`; we could, then, do a readlink on both `/proc/./self/exe` and `/proc/self/exe` and see if they differ, but the whole point of this PR is to avoid flakiness during upgrades (which often involve renaming or replacing files), and that sort of racy thing will end up as a bug report of "we had one `tsh ssh` failure but we can't replicate it" as some automatic upgrade thing was running in the background or something like that.

`exe` is not the only special case, however: `maps`, `stat` and `auxv` are also special cased in qemu (and have been for a really long time). The initial workaround I had involved opening `/proc/self/stat` and `/proc/./self/stat` and checking if their `st_dev` and `st_ino` matched, but while I was doing that I noticed that the real procfs file didn't report a real size while the qemu one (being a real file) did. I'd say that we can rely on the emulation of `/proc/self/stat` not getting more realistic before execve on `/proc/self/exe` starts working as intended - besides, this is a best effort thing anyway.

In fact, I seriously considered not caring about qemu-user at all (as plenty of reexec libraries seem to do, like [`github.com/docker/docker/pkg/reexec`](https://github.com/moby/moby/tree/master/pkg/reexec), but we have plenty of people using arm macs, and x86_64 docker containers run under qemu-user there, so...

Do you think I need to drop all of this in a comment in the code, or is the PR discussion good enough?

Reply:
Swapped out the test for a much simpler one 😅

@nklaassen can you PTAL?

Reply:
I'm going to trust you on all the `qemu` details, was just wondering if you found that check on the web somewhere we could link to or you came up with it.

But I'm really wondering if we can invert this condition? Instead of using `/proc/self/exe` unless (some mildly hacky check for qemu), we could only use `/proc/self/exe` if we can actually `Open` it and read a couple bytes or something, whatever is a sufficient check. Do you think that's feasible and could be safer in case some other random system breaks `/proc/self/exe`? It might properly handle more cases of qemu as well and not need a change if they ever fix this.

Reply:
Two problems with that:
- in the situation that we mostly care about (amd64 docker container on arm64 macos), because the executable is passed to qemu through the `AT_EXECFD` mechanism, a bug causes `open("/proc/self/exe")` to return the file descriptor number for the program at the time qemu was launched - in the best case scenario, it's not an open file descriptor and we get an error when we try to use it or when we close it, and in the worst case scenario it's a completely unrelated open file and we end up screwing with whatever was using it

- you might not have read permissions on a program you're running, technically

The alternate check I wrote at 51a376599242d60c35995da517e45df02f8f4562 is much easier to reason about, however.

Reply:
Oh and I don't believe there is a forward-compatible fix for this, the behavior is just too strange ☹️

Reply:
good points, thanks for taking the time to explain, your current solution LGTM then

Reply:
Got a forward-compatible fix in b02107d73bdcd4ad1dd04af3afe2b70c07940803